### PR TITLE
mycelium: init at 0.4.2

### PR DESCRIPTION
**Pull request by MatthewCroughan**
## Description of changes

Adds https://github.com/threefoldtech/mycelium to Nixpkgs at 0.4.2

#thaigersprint

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by flokli — CHANGES_REQUESTED** [review]
Can you bump this to 0.4.0?

File: pkgs/by-name/my/mycelium/package.nix [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    darwin.apple_sdk.frameworks.Security
+  ];
+
+  meta = with lib; {
+    description = "";
+    homepage = "https://github.com/threefoldtech/mycelium";
+    changelog = "https://github.com/threefoldtech/mycelium/blob/${src.rev}/CHANGELOG.md";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ matthewcroughan ];
+    mainProgram = "mycelium";
```
Comment (flokli) [inline pkgs/by-name/my/mycelium/package.nix:36]: What platforms did you try with? At least an impure `cargo build` on aarch64 did not work for me, see https://github.com/threefoldtech/mycelium/issues/132.

Comment (MatthewCroughan) [reply pkgs/by-name/my/mycelium/package.nix:36]: @flokli only on x86_64-linux, on aarch64-linu I get the following:

```
mycelium>    Compiling futures-channel v0.3.30
mycelium> error: linker `aarch64-linux-gnu-gcc` not found
mycelium>   |
mycelium>   = note: No such file or directory (os error 2)
mycelium> error: could not compile `proc-macro2` (build script) due to previous error
mycelium> warning: build failed, waiting for other jobs to finish...
mycelium> error: could not compile `serde` (build script) due to previous error
mycelium> error: could not compile `thiserror` (build script) due to previous error
mycelium> error: could not compile `libc` (build script) due to previous error
mycelium> error: could not compile `libc` (build script) due to previous error
mycelium> error: could not compile `network-interface` (build s
cript) due to previous error
mycelium> error: could not compile `anyhow` (build script) due to previous error
mycelium> error: could not compile `serde_json` (build script) due to previous error
mycelium> error: could not compile `async-trait` (build script) due to previous error
mycelium> error: could not compile `rustls` (build script) due to previous error
mycelium> error: could not compile `semver` (build script) due to previous error
mycelium> error: could not compile `rustix` (build script) due to previous error
mycelium> error: could not compile `httparse` (build script) due to previous error
mycelium> error: could not compile `rustversion` (build script) due to previous error
mycelium> error: could not compile `typenum` (build script) due to previous error
mycelium> error: could not compile `paste` (build script) due to previous error
```

Comment (flokli) [reply pkgs/by-name/my/mycelium/package.nix:36]: Yeah, that's expected - reported in https://github.com/threefoldtech/mycelium/issues/132 and fixed in https://github.com/threefoldtech/mycelium/commit/cce1cc6c209b7b41ad4b481da28d5d42dee84c06.

With this, I can even cross-compile a `mycelium.exe` - though I don't know if it works or not :laughing:

Comment (flokli) [reply pkgs/by-name/my/mycelium/package.nix:36]: Can you add x86_64-linux? That way we should get binaries built by hydra:

```
### `platforms` {#var-meta-platforms}

The list of Nix platform types on which the package is supported. Hydra builds packages according to the platform specified. If no platform is specified, the package does not have prebuilt binaries. An example is:
```

Comment (MatthewCroughan) [reply pkgs/by-name/my/mycelium/package.nix:36]: Will wait until 0.4.1

Comment (flokli) [reply pkgs/by-name/my/mycelium/package.nix:36]: Please add `platforms = [ "aarch64-darwin" "aarch64-linux" "x86_64-linux" ];`.

File: pkgs/by-name/my/mycelium/package.nix [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  ];
+
+  meta = with lib; {
+    description = "End-2-end encrypted IPv6 overlay network";
+    homepage = "https://github.com/threefoldtech/mycelium";
+    changelog = "https://github.com/threefoldtech/mycelium/blob/${src.rev}/CHANGELOG.md";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ matthewcroughan ];
+    mainProgram = "mycelium";
+  };
```
Comment (flokli) [inline pkgs/by-name/my/mycelium/package.nix:37]:
```suggestion
    platforms = [ "aarch64-darwin" "aarch64-linux" "x86_64-linux" ];
  };
```

**Review by flokli — APPROVED** [review]

**Comment by MatthewCroughan** [comment]
Final bump to 0.4.2 due to https://github.com/threefoldtech/mycelium/issues/140

**Pull request merged**